# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# PageSize is of type int unsigned so the result of buffer_offset%PageSize was 32b wide which then
# gets concatenates with host_address[31:0] bits. This writes to a non-existent storage in a 1MB
# Flash.

diff --git a/src/spi_host/model/spiflash.sv b/src/spi_host/model/spiflash.sv
index 412eff5..52eb0a0 100644
--- a/src/spi_host/model/spiflash.sv
+++ b/src/spi_host/model/spiflash.sv
@@ -595,7 +595,7 @@ program spiflash #(
     // Write data into storage
     buffer_offset = host_address[7:0];
     repeat (page_buffer_size) begin
-      storage[{host_address[31:8], buffer_offset%PageSize}] = page_buffer[buffer_offset%256];
+      storage[{host_address[31:8], 8'(buffer_offset%PageSize)}] = page_buffer[buffer_offset%256];
       buffer_offset = (buffer_offset+1) % PageSize;
     end
